rfc/manifest-validation.md

```diff
@@ -0,0 +1,451 @@
+    const report: Violation[] = [];
+    for (const manifest of manifests) {
+      const out = execSync(`datree test --output json ${manifest}`);
+      report.push(...parseOutput(out));
+    }
+    return report;
+  }
+}
+```
+
+During cdk8s synth, the CLI will look at the validations config property, download it from
+the remote source if necessary, and proceed to load and execute the plugins.
+
+Optionally, we could also provide a cdk8s validate command that can operate independently from synthesis.
+Instead of synthesizing manifests, it can accept them as a command line argument. This command can then
+be used in processes that don’t have access to the original source code, and operate solely on the output
+of cdk8s synth. One such example can be the deploy time validation gate we mentioned in the beginning.
+
+_FAQ_
+
+*Q*. Who is responsible for datree CLI to exist on the system? the user? the plugin? cdk8s?
+*A. *This would be for the plugin to decide. If embedding the tool inside an npm package is viable,
+they can do that. If not, the plugin will probably need to provide installation instructions.
+
+*Q*. Should cdk8s synth fail if the validation results in violations?
+*A.* Intuitively I’d say yes, but in order to debug the violation, the manifest itself will most likely be needed.
+So we will need to generate it, even if the validation fails. We should probably still exist with non-zero though, to indicate an error.
+
+### Construct Aware
+
+The ability to traceback to constructs world, will be enabled by adding a few annotations to the manifests generated
+by cdk8s. More concretely, we add cdk8s.io/construct.path and cdk8s.io/construct.traceback annotations:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: chart-deployment-c8a3b439
+  annotations:
+    cdk8s.io/construct.path: Chart/Deployment
+    cdk8s.io/construct.traceback:
+      - deployment.ts:25
+      - main.ts:15
+...
+...
+```
+
+This should be fairly easy to do by adding this logic in our base
+[ApiObject](https://github.com/cdk8s-team/cdk8s-core/blob/2.x/src/api-object.ts) class.
+Once we have it, since the violating resource is specified in the report, we can extract
+both the traceback and path from these annotations, to produce the final report we show to users.
+
+### Future Possibilities
+
+#### Advanced in-memory validations
+
+The mechanism described here suggests the validations are executed by the CLI. We could however, run
+the validations in the framework as well, as part of App.synth. This is somewhat more involved, but
+can allow writing more advanced plugins and validations, since they can have access to the
+entire in-memory construct tree.
+
+#### As-you-type validations
+
+The furthest left we can go, is providing immediate feedback as soon as a developer types a misconfiguration in their manifest.
+This can be achieved by editor plugins and can be extremely useful (but also very annoying if done poorly).
```

Comment:
Another way of shifting left could be if the validators had a "fix" capability, like eslint, that automatically updated your source code on error. For example, if the validator requires that every container image has a digest tag -- then it could look up the digest from a trustworthy source and automatically add it to your source code.